Improve sourcegraph release process docs (#2059)
sourcegraph/handbook: content/departments/product-engineering/engineering/process/releases/index.md

## Release policies

> > As of Jan 31, 2022, the [Delivery] team owns the release process. The role of Release Captain is referring to a member of the [Delivery] team. However, this may change in the future when we transit to a rotating base Release Captain.

Reviewer: ```suggestion
> As of Jan 31, 2022, the [Delivery] team owns the release process. The role of Release Captain is referring to a member of the [Delivery] team. However, this may change in the future when we transition to a rotation-based Release Captain.
```
I kinda don't want to fix this, because the original sentence makes me imagine something like this:
![hamster](https://user-images.githubusercontent.com/91073224/152027376-d9dbca27-40a3-47b4-9249-898a22d07606.gif)

Author: 😂English is hard

### Release tooling

> The [Sourcegraph release tool] is used to generate releases as associated materials (such as tracking issues).

Reviewer 2: Why remove this link? At the very least it should point to the source code

Author: > Why remove this link? At the very least it should point to the source code

It's some fancy markdown anchor link syntax, we can reuse the same anchor link to keep things DRY. 

https://github.com/sourcegraph/handbook/pull/2059/files#diff-23bdf5adea38c4be7b98df79bb9f1b1959d5380a25ad3e269abccb12f91fb0f2R249

```md
[sourcegraph release tool] is great

[sourcegraph release tool] is awesome

[sourcegraph release tool]: ../../tools/release.md
```

Most issues are non-blocking. Fixes to non-blocking issues can be fixed in `main` by the code owner who can then `git cherry-pick` those commits into the release branch with the approval of the release captain. Alternatively, broken features can be reverted out of the release branch or disabled via feature flags if they aren't ready or are too buggy.

## Patch release process

> > As of Jan 31, 2022, the Delivery team still owns the release process hence the role of Release Captain, `us` in the process below is always refering to the Delivery team. However, this may change in the future when we transit to a rotation base Release Captain.

Reviewer: There are a few typos / grammatical errors in the doc - probably worth running a spellchecker against your changes.

> ### 1) Kickstart a patch release

Reviewer: ```suggestion
### 1) Kickstart a patch release using the release tool
```

Reviewer: Small suggestion - I'm hoping that we can make it so that a release captain can scan the headings to get a reminder of what to do, and only dive into the details if necessary.

CE or the products team will start requesting a patch release by submitting a [patch release request]. The patch release to [revert poor onboarding UX change] is a good example.

> Upon the Delivery team or the Release Captain receives the patch release request and we have decided to rollout a new patch release, the Release Captain should follow the instruction in the [patch release request] issue to kickstart the patch release process.

Reviewer: Maybe we should add another section detailing how to decide whether to accept the patch - who gets to decide? Who needs to be part of the conversation, and who gets notified? What's the timeframe?

I don't think we have good answers - pretty much the release captain says 'yes' and immediately works on the patch. CS generally gets looped in by either the dev requesting the patch or the release captain, since a patch almost always causes additional work for them.

The instruction is located at the bottom of the [patch release request]. At a high level, it contains the following

- Open a PR to update [release-config.jsonc] config file with the upcoming release information and merge the PR
- Utilize the [Sourcegraph release tool] to initialize the next step

### 2) Patch release tracking issue

The [Sourcegraph release tool] creates a patch release tracking issue which contains a list of action items the Release Captain has to perform. [3.36.2 patch release tracking issue] is a good example.

The Release Captain should review and follow the instruction in the patch release tracking release for next steps. At a high level, it contains the followin steps

> #### 2.1) Cherry-pick patch commits from `main` into the release branch

Reviewer: I wonder how valuable it is to include these steps - they're already included in the tracking issue itself, and it means that if we change that issue this list can get out of sync.

Maybe instead, we can just summarize the intent of the process? Build a release candidate, verify it passes CI, publish final image and update documentation/references.

Reviewer 2: +1 - the release issue documents this, and the issue template is already available in the handbook

Author: good point, also the handbook and the issue template could eventually get out of sync.

I am removing all specific steps here and move some nice to have notes (the one about how to merge the batch change) into the release template.

Wait for CI passes on the releas branches listed.

#### 2.2) Tag a release candiate build

Wait for CI passes for the new release candidate build.

#### 2.3) Tag the final patch release build

Wait for CI passes for the new patch tag.

#### 2.4) Update image tag acroos repos with batch change

This will make neccessary code changes across multiple repositories. [3.36.2 release batch change](https://k8s.sgdev.org/organizations/sourcegraph/batch-changes/release-sourcegraph-3.36.2?visible=6) is a good example.

Changesets from the release batch change have to be merged following a specific order.

1. sourcegraph/deploy-sourcegraph-digitalocean
1. sourcegraph/deploy-sourcegraph-aws
1. sourcegraph/deploy-sourcegraph-docker
1. sourcegraph/deploy-sourcegraph
1. sourcegraph/about
1. sourcegraph/sourcegraph

> Unless there's specific instruction in the changeset (Pull Request on GitHub), the Release Captain can merge them with confident provided that CI checks pass.

Reviewer: It might be better to reword this as an active sentence - confirm that CI checks have passed and all manual steps detailed in the PR have been addressed

The changeset against [sourcegraph/sourcegraph] requires some manual changes to the changelog. Follow the instruction in the Pull Request description.

Special attention is required to the changeset against [sourcegraph/deploy-sourcegraph-docker]. The changeset in [sourcegraph/deploy-sourcegraph-docker] also contains the release of our `pure-docker` deployment method and it is error prone (you need to manually resolved an expected merge conflict). Follow the instruction provided in the Pull Request description.

#### 2.5) Wrapping up

Follow the instruction from post-release in the patch release tracking issue to wrap up the patch release

### 3) Revisit patch request issue

Now it's a good time to go back to the original [patch release request] and close it.

> ## Minor release process

Reviewer: Totally silly, but in my mind "Minor release" should go before "Patch release" in the doc 😄

Author: lol, it makes sense. I just happened to write up the patch release first. I'll move them around.

### 1) Start a minor release

Major and minor releases are released on a fixed scheule, see [when we release](#when-we-release).

### 2) Minor release tracking issue

The tracking issue for the current minor release is created as a part of the post-release step from the previous minor release. [3.36.0 release tracking issue] is a good example of a minor release tracking issue.

> The Release Captain should review and follow the instruction in the patch release tracking release for next steps. At a high level, it contains the followin steps

Reviewer: Again, I think it could be worthwhile to replace the specific steps with a summary of the intent.

#### 2.1) Open a PR to pdate [release-config.jsonc]

Merge this PR prior proceeding to the next step

#### 2.2) Create a new release branch

For example, `3.36`.

#### 2.3) Tag a release candidate build

> Wait for CI passes for the new release candidate build.

Reviewer 3: Not a definite thing, but my question from the outside perspective, is "Why do I need to wait for CI to pass"? I'm wondering if it's worth putting that type of information in here? 🤷

Author: LOL this is my impression too, like the Release Captain has to wait for multiple CI builds and each takes 30+min to run

- the cherry-pick one on the release branch
- release candidate tag build
- final release tag build

so much time is wasted. 

Maybe we can only wait for one of the CI runs to pass?

Most issues are non-blocking. Fixes to non-blocking issues can be fixed in `main` by the code owner who can then `git cherry-pick` those commits into the release branch with the approval of the release captain. Alternatively, broken features can be reverted out of the release branch or disabled via feature flags if they aren't ready or are too buggy.

## Minor release process

### 1) Start a minor release

Major and minor releases are released on a fixed scheule, see [when we release](#when-we-release).

### 2) Minor release tracking issue

> The tracking issue for the current minor release is created as a part of the post-release step from the previous minor release. [3.36.0 release tracking issue] is a good example of a minor release tracking issue.

Reviewer 2: Similar: We should just link to the template here, instead of a potentially outdated example

> The Release Captain should review and follow the instruction in the patch release tracking release for the next steps. At a high level, it includes the following steps:

Reviewer: ```suggestion
The Release Captain should review and follow the instruction in the release tracking release for the next steps. At a high level, it includes the following steps:
```

- Build a release candidate and verify CI passes
- Publish final images
- Update documentation
- Update references of the image tag to the new version using batch change

### 3) Wrapping up

> Follow the instruction from post-release in the release tracking issue to wrap up the release.

Reviewer: ```suggestion
Follow the instruction from post-release in the release tracking issue to wrap up the release and schedule the next minor release.
```

## Patch release process

### 1) Kickstart a patch release using the [Sourcegraph release tool]

CE or the products team will start requesting a patch release by submitting a [patch release request]. The patch release to [revert poor onboarding UX change] is a good example.

Upon the Delivery team or the Release Captain receives the patch release request and we have decided to rollout a new patch release, the Release Captain should follow the instruction in the [patch release request] issue to kickstart the patch release process. The instruction is located at the bottom of the [patch release request].

### 2) Patch release tracking issue

> The [Sourcegraph release tool] creates a patch release tracking issue which contains a list of action items the Release Captain has to perform. [3.36.2 patch release tracking issue] is a good example.

Reviewer 2: We should just link to the template here, instead of a potentially outdated example

> - We haven't built the automated testing and update infrastructure to make continuous customer releases reliable.

Reviewer 3: Is this true? I think it's actually more nuanced than this. Our customers usually don't have the capacity to upgrade continuously. It can take a customer months (or years) to get to the next incremental release depending on their internal processes. A lot of our customers have rigorous change controls in-place.

I agree that our automated testing could use some work, but definitely think the issue at hand is a little more complicated 😆 

I know you didn't make this change, just saw it and felt something about this line!

Author: I am a bit confused.

Are you saying we do have some `automated testing and update infrastructure` for releases? 

About enabling customers to continuously update, the biggest blocker is 

> We don't think customers would be comfortable with a continuously updated service running on their own infrastructure, for security and stability reasons.

If we throw this away, and we do want to CD for on-prem. I saw Dan is talking about giving [Replicated](https://www.replicated.com/) another try. This could potentially allow us to have an automated release process for customers.
